File: LXToolKit/LXToolKit2/Category/LXPlatformInfo.m
//
//  LXPlatformInfo.m
//  Vaffle_demo
//
//  Created by DamonJow on 2018/11/1.
//  Copyright © 2018 DamonJow. All rights reserved.
//

#import "LXPlatformInfo.h"

#include <sys/types.h>
#include <sys/sysctl.h>

@interface LXPlatformInfo () {
}
/** platformInfo */
@property(nonatomic, retain)NSDictionary *platform;
@end
@implementation LXPlatformInfo

+ (instancetype)sharedInstance {
    static LXPlatformInfo *_sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedInstance = [[LXPlatformInfo alloc]init];
    });
    return _sharedInstance;
}

/// 获得设备型号
+ (NSString *)getCurrentDeviceModel {
    return [[LXPlatformInfo sharedInstance] getCurrentDeviceModel];
}
/// 获得设备型号
- (NSString *)getCurrentDeviceModel {
    int mib[2];
    size_t len;
    char *machine;

    mib[0] = CTL_HW;
    mib[1] = HW_MACHINE;
    sysctl(mib, 2, NULL, &len, NULL, 0);
    machine = malloc(len);
    sysctl(mib, 2, machine, &len, NULL, 0);

    NSString *platform = [NSString stringWithCString:machine encoding:NSASCIIStringEncoding];
    free(machine);

    return self.platform[platform];
}

#pragma mark - platform
- (NSDictionary *)platform {
    if(!_platform){
        _platform = @{
                                    @"i386" : @"iPhone Simulator",
                                    @"x86_64" : @"iPhone Simulator",

                                    /// iPhone
                                    @"iPhone1,1" : @"iPhone 2G (A1203)",
                                    @"iPhone1,2" : @"iPhone 3G (A1241/A1324)",
                                    @"iPhone2,1" : @"iPhone 3GS (A1303/A1325)",
                                    @"iPhone3,1" : @"iPhone 4 (A1332)",
                                    @"iPhone3,2" : @"iPhone 4 (A1332)",
                                    @"iPhone3,3" : @"iPhone 4 (A1349)",
                                    @"iPhone4,1" : @"iPhone 4S (A1387/A1431)",
                                    @"iPhone5,1" : @"iPhone 5 (A1428)",
                                    @"iPhone5,2" : @"iPhone 5 (A1429/A1442)",
                                    @"iPhone5,3" : @"iPhone 5c (A1456/A1532)",
                                    @"iPhone5,4" : @"iPhone 5c (A1507/A1516/A1526/A1529)",
                                    @"iPhone6,1" : @"iPhone 5s (A1453/A1533)",
                                    @"iPhone6,2" : @"iPhone 5s (A1457/A1518/A1528/A1530)",
                                    @"iPhone7,1" : @"iPhone 6 Plus (A1522/A1524)",
                                    @"iPhone7,2" : @"iPhone 6 (A1549/A1586)",
                                    @"iPhone8,1" : @"iPhone 6S",
                                    @"iPhone8,2" : @"iPhone 6S Plus",
                                    @"iPhone8,3" : @"iPhone SE",
                                    @"iPhone8,4" : @"iPhone SE",
                                    @"iPhone9,1" : @"iPhone 7",
                                    @"iPhone9,2" : @"iPhone 7 Plus",
                                    @"iPhone10.1" : @"iPhone 8",
                                    @"iPhone10,4" : @"iPhone 8",
                                    @"iPhone10.2" : @"iPhone 8 Plus",
                                    @"iPhone10,5" : @"iPhone 8 Plus",
                                    @"iPhone10,3" : @"iPhone X",
                                    @"iPhone10,6" : @"iPhone X",

                                    /// iPod
                                    @"iPod1,1" : @"iPod Touch 1G",
                                    @"iPod2,1" : @"iPod Touch 2G",
                                    @"iPod3,1" : @"iPod Touch 3G",
                                    @"iPod4,1" : @"iPod Touch 4G",
                                    @"iPod5,1" : @"iPod Touch 5G",
                                    @"iPod7,1" : @"iPod Touch 6G",

                                    /// iPad
                                    @"iPad1,1" : @"iPad",
                                    @"iPad2,1" : @"iPad 2 (WiFi)",
                                    @"iPad2,2" : @"iPad 2 (GSM)",
                                    @"iPad2,3" : @"iPad 2 (CDMA)",
                                    @"iPad2,4" : @"iPad 2 (32nm)",
                                    @"iPad3,1" : @"iPad 3 (WiFi)",
                                    @"iPad3,2" : @"iPad 3(CDMA)",
                                    @"iPad3,3" : @"iPad 3(4G)",
                                    @"iPad3,4" : @"iPad 4 (WiFi)",
                                    @"iPad3,5" : @"iPad 4 (4G)",
                                    @"iPad3,6" : @"iPad 4 (CDMA)",

                                    /// iPad Air
                                    @"iPad4,1" : @"iPad Air",
                                    @"iPad4,2" : @"iPad Air",
                                    @"iPad4,3" : @"iPad Air",
                                    @"iPad5,3" : @"iPad Air 2",
                                    @"iPad5,4" : @"iPad Air 2",

                                    /// iPad mini
                                    @"iPad2,5" : @"iPad Mini (WiFi)",
                                    @"iPad2,6" : @"iPad Mini (GSM)",
                                    @"iPad2,7" : @"iPad Mini (CDMA)",
                                    @"iPad4,4" : @"iPad mini 2",
                                    @"iPad4,5" : @"iPad mini 2",
                                    @"iPad4,6" : @"iPad mini 2",
                                    @"iPad4,7" : @"iPad mini 3",
                                    @"iPad4,8" : @"iPad mini 3",
                                    @"iPad4,9" : @"iPad mini 3",
                                    @"iPad5,1" : @"iPad mini 4",
                                    @"iPad5,2" : @"iPad mini 5",
                                    };
    }
    return _platform;
}
@end

File: LXToolKit/LXToolKit2/Category/LXPlatformInfo.swift
//
//  LXPlatformInfo.swift
//  Vaffle_demo
//
//  Created by DamonJow on 2018/11/1.
//  Copyright © 2018 DamonJow. All rights reserved.
//
import UIKit
import Foundation

public extension UIDevice {

    /// let modelName = UIDevice.currentDevice().ts_platform
    var xl_platform: String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8 , value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }

        switch identifier {
        //iPhone
        case "iPhone1,1": return "iPhone 1G"
        case "iPhone1,2": return "iPhone 3G"
        case "iPhone2,1": return "iPhone 3GS"
        case "iPhone3,1": return "iPhone 4 (GSM)"
        case "iPhone3,3": return "iPhone 4 (CDMA)"
        case "iPhone4,1": return "iPhone 4S"
        case "iPhone5,1": return "iPhone 5 (GSM)"
        case "iPhone5,2": return "iPhone 5 (CDMA)"
        case "iPhone5,3": return "iPhone 5c"
        case "iPhone5,4": return "iPhone 5c"
        case "iPhone6,1": return "iPhone 5s"
        case "iPhone6,2": return "iPhone 5s"
        case "iPhone7,1": return "iPhone 6 Plus"
        case "iPhone7,2": return "iPhone 6"
        case "iPhone8,1": return "iPhone 6s"
        case "iPhone8,2": return "iPhone 6s Plus"
        case "iPhone8,4": return "iPhone SE"
        case "iPhone9,1": return "iPhone 7"
        case "iPhone9,3": return "iPhone 7"
        case "iPhone9,2": return "iPhone 7 Plus"
        case "iPhone9,4": return "iPhone 7 Plus"
        case "iPhone10,1": return "iPhone 8"
        case "iPhone10,4": return "iPhone 8"
        case "iPhone10,2": return "iPhone 8 Plus"
        case "iPhone10,5": return "iPhone 8 Plus"
        case "iPhone10,3": return "iPhone X"
        case "iPhone10,6": return "iPhone X"
        case "iPhone11,8": return "iPhone XR"
        case "iPhone11,2": return "iPhone XS"
        case "iPhone11,4": return "iPhone XS MAX"
        case "iPhone11,6": return "iPhone XS MAX"
        case "iPhone12,1": return "iPhone11"
        case "iPhone12,3": return "iPhone11 Pro"
        case "iPhone12,5": return "iPhone11 Pro Max"
        //iPod
        case "iPod1,1": return "iPod Touch 1G"
        case "iPod2,1": return "iPod Touch 2G"
        case "iPod3,1": return "iPod Touch 3G"
        case "iPod4,1": return "iPod Touch 4G"
        case "iPod5,1": return "iPod Touch 5G"
        case "iPod7,1": return "iPod Touch 6G"
        //iPad
        case "iPad1,1": return "iPad"
        case "iPad1,2": return "iPad 3G"
        case "iPad2,1": return "iPad 2 (WiFi)"
        case "iPad2,2": return "iPad 2 (GSM)"
        case "iPad2,3": return "iPad 2 (CDMA)"
        case "iPad2,4": return "iPad 2 (WiFi)"
        case "iPad2,5": return "iPad Mini (WiFi)"
        case "iPad2,6": return "iPad Mini (GSM)"
        case "iPad2,7": return "iPad Mini (CDMA)"
        case "iPad3,1": return "iPad 3 (WiFi)"
        case "iPad3,2": return "iPad 3 (CDMA)"
        case "iPad3,3": return "iPad 3 (GSM)"
        case "iPad3,4": return "iPad 4 (WiFi)"
        case "iPad3,5": return "iPad 4 (GSM)"
        case "iPad3,6": return "iPad 4 (CDMA)"
        case "iPad4,1": return "iPad Air (WiFi)"
        case "iPad4,2": return "iPad Air (GSM)"
        case "iPad4,3": return "iPad Air (CDMA)"
        case "iPad4,4": return "iPad Mini Retina (WiFi)"
        case "iPad4,5": return "iPad Mini Retina (Cellular)"
        case "iPad4,6": return "iPad mini Retina (China)"
        case "iPad4,7": return "iPad Mini 3 (WiFi)"
        case "iPad4,8": return "iPad Mini 3 (Cellular)"
        case "iPad4,9": return "iPad Mini 3 (Cellular)"
        case "iPad5,1": return "iPad Mini 4 (WiFi)"
        case "iPad5,2": return "iPad Mini 4 (Cellular)"
        case "iPad5,3": return "iPad Air 2 (WiFi)"
        case "iPad5,4": return "iPad Air 2 (Cellular)"
        case "iPad6,3": return "iPad Pro 9.7-inch (WiFi)"
        case "iPad6,4": return "iPad Pro 9.7-inch (Cellular)"
        case "iPad6,7": return "iPad Pro 12.9-inch (WiFi)"
        case "iPad6,8": return "iPad Pro 12.9-inch (Cellular)"
        case "iPad6,11": return "iPad (2017)"
        case "iPad6,12": return "iPad (2017)"
        case "iPad7,1": return "iPad Pro 2nd Gen (WiFi)"
        case "iPad7,2": return "iPad Pro 2nd Gen (WiFi+Cellular)"
        case "iPad7,3": return "iPad Pro 10.5-inch"
        case "iPad7,4": return "iPad Pro 10.5-inch"
        case "iPad7,5": return "iPad 6th Gen (WiFi)"
        case "iPad7,6": return "iPad 6th Gen (WiFi+Cellular)"
        case "iPad8,1": return "iPad Pro 3rd Gen (11 inch, WiFi)"
        case "iPad8,2": return "iPad Pro 3rd Gen (11 inch, 1TB, WiFi)"
        case "iPad8,3": return "iPad Pro 3rd Gen (11 inch, WiFi+Cellular)"
        case "iPad8,4": return "iPad Pro 3rd Gen (11 inch, 1TB, WiFi+Cellular)"
        case "iPad8,5": return "iPad Pro 3rd Gen (12.9 inch, WiFi)"
        case "iPad8,6": return "iPad Pro 3rd Gen (12.9 inch, 1TB, WiFi)"
        case "iPad8,7": return "iPad Pro 3rd Gen (12.9 inch, WiFi+Cellular)"
        case "iPad8,8": return "iPad Pro 3rd Gen (12.9 inch, 1TB, WiFi+Cellular)"
        //AppleTV
        case "AppleTV5,3": return "Apple TV"
        //Simulator
        case "i386", "x86_64": return "Simulator"
        default: return identifier
        }
    }
}
